blackjax/inference/base.py

```diff
@@ -0,0 +1,200 @@
+        after which we consider the transition to be divergent.
+
+    Returns
+    -------
+    A kernel that moves the chain by one step when called.
+
+    References
+    ----------
+    .. [1]: Duane, Simon, et al. "Hybrid monte carlo." Physics letters B
+            195.2 (1987): 216-222.
+    .. [2]: Neal, Radford M. "An improved acceptance procedure for the
+            hybrid Monte Carlo algorithm." Journal of Computational Physics 111.1
+            (1994): 194-203.
+    .. [3]: Betancourt, Michael. "A conceptual introduction to
+            Hamiltonian Monte Carlo." arXiv preprint arXiv:1701.02434 (2018).
+    .. [4]: "Hamiltonian Mechanics", Wikipedia.
+            https://en.wikipedia.org/wiki/Hamiltonian_mechanics#Deriving_Hamilton's_equations
+    .. [5]: Betancourt, Michael, et al. "The geometric foundations
+            of hamiltonian monte carlo." Bernoulli 23.4A (2017): 2257-2298.
+
+    """
+
+    def kernel(
+        rng_key: jax.random.PRNGKey, state: HMCState
+    ) -> Tuple[HMCState, HMCInfo]:
+        """Moves the chain by one step using the Hamiltonian dynamics.
+
+        Parameters
+        ----------
+        rng_key:
+           The pseudo-random number generator key used to generate random numbers.
+        state:
+            The current state of the chain: position, log-probability and gradient
+            of the log-probability.
+
+        Returns
+        -------
+        The next state of the chain and additional information about the current step.
+        """
+        key_momentum, key_integrator, key_accept = jax.random.split(rng_key, 3)
+
+        position, potential_energy, potential_energy_grad = state
+        momentum = momentum_generator(key_momentum, position)
+        energy = potential_energy + kinetic_energy(momentum, position)
+
+        proposal, proposal_info = proposal_generator(
+            key_integrator, HMCProposalState(position, momentum, potential_energy_grad)
+        )
+        new_position, new_momentum, new_potential_energy_grad = proposal
+
+        flipped_momentum = jax.tree_util.tree_multimap(lambda m: -1.0 * m, new_momentum)
+        new_potential_energy = potential_fn(new_position)
+        new_energy = new_potential_energy + kinetic_energy(
+            flipped_momentum, new_position
+        )
+        new_state = HMCState(
+            new_position, new_potential_energy, new_potential_energy_grad
+        )
+
+        delta_energy = energy - new_energy
+        delta_energy = jnp.where(jnp.isnan(delta_energy), -jnp.inf, delta_energy)
+        is_divergent = jnp.abs(delta_energy) > divergence_threshold
+
+        p_accept = jnp.clip(jnp.exp(delta_energy), a_max=1)
+        do_accept = jax.random.bernoulli(key_accept, p_accept)
+        accept_state = (
+            new_state,
+            HMCInfo(
+                p_accept,
+                True,
+                is_divergent,
+                new_energy,
+                proposal,
```

Comment:
nit: If we store `potential_energy` in `HMCProposalState`, we can probably have a single `HMCState` for both representing `(q, p, potential_energy, potential_energy_grad)` and not have to store `proposal` separately in `HMCInfo`. Please disregard my comment if there are other reasons for this separation.

Reply:
The `proposal` in `HMCInfo` is the last state of the trajectory. `new_state` is what I now think is what should be called the proposal. I return the last state of the trajectory in `HMCInfo` because it can be useful to diagnose when there are divergences or the acceptance rate is very low.

Reply:
I see, that makes sense. If a proposal is rejected, we have both the earlier state and the proposed state to debug why it was rejected.